### PR TITLE
[sec-approval/1] api: report secure revisions (bug 1539289)

### PR DESCRIPTION
**Pull request by mars-f**
Revision entities returned by the API publish a new "is_secure"
property.  Revisions for which this property is True contain
security-sensitive information and should be handled appropriately by
clients.

### CODE REVIEW

File: landoapi/api/transplants.py [hunk 1]
```diff
@@ -5,7 +5,7 @@
 import urllib.parse
 
 import kombu
-from connexion import problem, ProblemException
+from connexion import ProblemException, problem
```
Comment (smacleod) [inline landoapi/api/transplants.py]: Convention throughout lando-api is to ignore case when alphabetizing, please revert the changes to imports here and elsewhere.

File: landoapi/api/transplants.py [hunk 2]
```diff
@@ -171,7 +171,9 @@ def _assess_transplant_request(phab, landing_path):
     reviewers = {
         revision["phid"]: get_collated_reviewers(revision) for revision, _ in to_land
     }
+    is_secure = revision_is_secure(revision, phab)
 
+    # fmt: off
```
Comment (smacleod) [inline landoapi/api/transplants.py]: Why? This code seems pretty basic, I don't see why it warrants special formatting?

File: landoapi/revisions.py [hunk 3]
```diff
@@ -118,3 +119,26 @@ def check_author_planned_changes(*, revision, **kwargs):
...
+
+    Such revisions should be handled according to the Security Bug Approval Process.
+    See https://wiki.mozilla.org/Security/Bug_Approval_Process.
+
+    Args:
+        revision: The 'data' element from a Phabricator API response containing
+            revision data.
+        phabclient: A PhabricatorClient instance.
+    """
+    secure_project_phid = get_secure_project_phid(phabclient)
```
Comment (smacleod) [inline landoapi/revisions.py]: Every time this is called we go across the network to the redis cache at minimum, I'd much rather we leave things where the phid is grabbed once as part of the request and then passed around. Making this take the `phid` rather than a full phabclient means it's a pure function with only basic types going in and out as well.

Particularly I like to avoid sending phabclients down into functions unless there's a ton of value, to avoid the creep we had before where everything was layered super deep and calls were being made to phabricator all over the place. The general way things are laid out now is to call to external systems in the top level request most of the time and then pass just basic python types around.

The help functions on phabclient, such as `expect` are static methods, which again are pure functions, so you can make use of them directly off of the class `PhabricatorClient.expect(...)` without needing the instantiated client.

File: landoapi/api/transplants.py [hunk 4]
```diff
@@ -180,8 +182,10 @@ def _assess_transplant_request(phab, landing_path):
         reviewers,
         users,
         projects,
-        get_secure_project_phid(phab),
+        is_secure,
```
Comment (smacleod) [inline landoapi/api/transplants.py]: You've changed the semantics here. Previously we were checking the security of each revision, but now this would check only a single revision and apply it's security status to all the revisions.

Please revert things to just pass the `phid` and let the warning system check each revision. Having the new `revision_is_secure()` function will still simplify the warning body a lot.

File: landoapi/revisions.py [hunk 5]
```diff
@@ -118,3 +119,26 @@ def check_author_planned_changes(*, revision, **kwargs):
...
+
+def revision_is_secure(revision, phabclient):
+    """Does the given revision contain security-sensitive data?
+
+    Such revisions should be handled according to the Security Bug Approval Process.
+    See https://wiki.mozilla.org/Security/Bug_Approval_Process.
+
+    Args:
+        revision: The 'data' element from a Phabricator API response containing
+            revision data.
```
Comment (smacleod) [inline landoapi/revisions.py]: This isn't completely correct, it specifically needs to be a `differential.revision.search` response for a revision, which included the `projects` attachment. We have a convention for pulling all the data for the stack in the one place, so that it has all we need and is standard, you could refer to the stack data stuff if you want this comment to require less updating.

File: tests/test_revisions.py [hunk 6]
```diff
@@ -65,3 +69,46 @@ def test_check_author_planned_changes_changes_planned(phabdouble):
         attachments={"reviewers": True, "reviewers-extra": True, "projects": True},
     )["data"][0]
     assert check_author_planned_changes(revision=revision) is not None
+
+
+def test_secure_api_flag_on_public_revision_is_false(client, phabdouble):
```
Comment (smacleod) [inline tests/test_revisions.py]: You should probably update one of the integrated stack tests to make something secure, as that would have caught the issue around assuming security based on a single revision's status.

File: landoapi/api/transplants.py (hunk 2 above)
Comment (mars-f) [reply landoapi/api/transplants.py]: This change was reverted with the old warning behaviour.

File: landoapi/revisions.py (hunk 3 above)
Comment (mars-f) [reply landoapi/revisions.py]: Changed.

File: landoapi/api/transplants.py (hunk 4 above)
Comment (mars-f) [reply landoapi/api/transplants.py]: I added an integration test to catch the bug and reverted the call chain to the old behaviour.

File: landoapi/revisions.py (hunk 5 above)
Comment (mars-f) [reply landoapi/revisions.py]: I used the docstring from another function with a "revision" data argument as a template to clarify this docstring.

File: tests/test_revisions.py (hunk 6 above)
Comment (mars-f) [reply tests/test_revisions.py]: Agreed.  I added an integrated transplant test to ensure the warnings are computed correctly and an integrated stack test to make sure the stack values are correct.

File: landoapi/api/stacks.py [hunk 7]
```diff
@@ -103,7 +104,6 @@ def get(revision_id):
             title, bug_id, accepted_reviewers, summary, revision_url
         )
         author_response = serialize_author(phab.expect(fields, "authorPHID"), users)
-
```
[the remark this answers is not in the document]
Comment (mars-f) [reply landoapi/api/stacks.py]: No idea what that's about.  Reverted.

File: landoapi/api/stacks.py [hunk 8]
```diff
@@ -126,6 +126,9 @@ def get(revision_id):
                 "diff": serialize_diff(diff),
                 "author": author_response,
                 "reviewers": serialize_reviewers(reviewers, users, projects, diff_phid),
+                "is_secure": revision_is_secure(
+                    revision, get_secure_project_phid(phab)
```
[the remark this answers is not in the document]
Comment (mars-f) [reply landoapi/api/stacks.py]: Fixed.

File: landoapi/revisions.py [hunk 9]
```diff
@@ -118,3 +118,26 @@ def check_author_planned_changes(*, revision, **kwargs):
...
+        secure_project_phid: The PHID of the Phabricator project used to tag
+            secure revisions.
+    """
+    revision_project_tags = PhabricatorClient.expect(
+        revision, "attachments", "projects", "projectPHIDs"
+    )
+    is_secure = secure_project_phid in revision_project_tags
+    logger.debug(
+        "revision is security-sensitive?",
+        extra={"value": is_secure, "revision": revision["id"]},
```
[the remark this answers is not in the document]
Comment (mars-f) [reply landoapi/revisions.py]: Agreed, removed.

File: landoapi/transplants.py [hunk 10]
```diff
@@ -242,17 +246,14 @@ def warning_revision_secure(*, revision, secure_project_phid, **kwargs):
     if secure_project_phid is None:
         return None
 
-    revision_project_tags = PhabricatorClient.expect(
-        revision, "attachments", "projects", "projectPHIDs"
-    )
-    if secure_project_phid not in revision_project_tags:
+    if revision_is_secure(revision, secure_project_phid):
```
[the remark this answers is not in the document]
Comment (mars-f) [reply landoapi/transplants.py]: Done.

File: tests/test_revisions.py [hunk 11]
```diff
@@ -65,3 +69,47 @@ def test_check_author_planned_changes_changes_planned(phabdouble):
...
+    revision = phabdouble.revision(projects=[secure_project])
+
+    response = client.get("/stacks/D{}".format(revision["id"]))
+
+    assert response.status_code == 200
+    response_revision = response.json["revisions"].pop()
+    assert response_revision["is_secure"]
+
+
+def test_public_revision_is_not_secure(app, phabdouble, secure_project):
```
[the remark this answers is not in the document]
Comment (mars-f) [reply tests/test_revisions.py]: It's leftover from an earlier version of this test.  Removed.

File: tests/test_revisions.py [hunk 12]
```diff
@@ -65,3 +69,47 @@ def test_check_author_planned_changes_changes_planned(phabdouble):
...
+    r = phabdouble.revision(projects=[public_project])
+    revision = phab.call_conduit(
+        "differential.revision.search",
+        constraints={"phids": [r["phid"]]},
+        attachments={"reviewers": True, "reviewers-extra": True, "projects": True},
+    )["data"].pop()
+    assert not revision_is_secure(revision, secure_project["phid"])
+
+
+def test_secure_revision_is_secure(app, phabdouble, secure_project):
```
[the remark this answers is not in the document]
Comment (mars-f) [reply tests/test_revisions.py]: It's leftover from an earlier version of this test.  Removed.

File: tests/test_transplants.py [hunk 13]
```diff
@@ -142,6 +142,46 @@ def test_dryrun_reviewers_warns(client, db, phabdouble, auth0_mock):
     assert response.json["confirmation_token"] is not None
 
 
+def test_dryrun_secure_revision_in_stack_warns(
```
[the remark this answers is not in the document]
Comment (mars-f) [reply tests/test_transplants.py]: I agree that this test can be removed, I think this test is narrow and unlikely to catch future errors.  An earlier version of this PR broke how warnings were calculated, however no tests failed.  The test was written to expose the bug and verify the fix.
